static/js/services/get-subject-summaries.js

```diff
@@ -0,0 +1,131 @@
+
+angular.module('inboxServices').factory('GetSubjectSummaries',
+  function(
+    $q,
+    DB
+  ) {
+
+    'use strict';
+    'ngInject';
+
+    var findSubjectId = function(response, id) {
+      var parent = _.find(response.rows, function(row) {
+        return id && row.key[1] === id.toString() || false;
+      });
+      return (parent && parent.id) || null;
+    };
+
+    var replaceReferencesWithIds = function(summaries, response) {
+      summaries.forEach(function(summary) {
+        if (summary.subject.type === 'reference' && summary.subject.value) {
+          var id = findSubjectId(response, summary.subject.value);
+          if (id) {
+            summary.subject = {
+              value: id,
+              type: 'id'
+            };
+          }
+        }
+      });
+
+      return summaries;
+    };
+
+    var findSubjectName = function(response, id) {
+      var parent = _.findWhere(response.rows, { id: id });
+      return (parent && parent.value && parent.value.name) || null;
+    };
+
+    var replaceIdsWithNames = function(summaries, response) {
+      summaries.forEach(function(summary) {
+        if (summary.subject && summary.subject.type === 'id' && summary.subject.value) {
+          var name = findSubjectName(response, summary.subject.value);
+          if (name) {
+            summary.subject = {
+              value: name,
+              type: 'name'
+            };
+          }
+        }
+      });
+      return summaries;
+    };
+
+    var processContactIds = function(summaries) {
+      var ids = _.uniq(_.compact(_.flatten(summaries.map(function(summary) {
+        if (summary.subject && summary.subject.type === 'id') {
+          return summary.subject.value;
+        }
+      }))));
+
+      if (!ids.length) {
+        return $q.resolve(summaries);
+      }
+
+      return DB()
+        .query('medic-client/doc_summaries_by_id', { keys: ids })
+        .then(function(response) {
+          return replaceIdsWithNames(summaries, response);
+        });
+    };
+
+
+    var validateSubjects = function(summaries) {
+      summaries.forEach(function(summary) {
+        if (!summary.subject) {
+          return;
+        }
+
+        summary.valid_subject = true;
+
+        if (!summary.subject.type) {
+          summary.subject.value = summary.contact || summary.from;
+        } else if (summary.subject.type !== 'name' || !summary.subject.value) {
+          summary.valid_subject = false;
+        }
+      });
+
+      return summaries;
+    };
+
+    var processReferences = function(summaries) {
+      var references = _.uniq(_.compact(_.flatten(summaries.map(function(summary) {
+        if (summary.subject && summary.subject.type === 'reference') {
+          return summary.subject.value;
+        }
+      }))));
+
+      if (!references.length) {
+        return $q.resolve(summaries);
+      }
+
+      references.forEach(function(reference, key) {
+        references[key] = ['shortcode', reference];
+      });
```

Comment:
Instead of looping through twice why not have the map on line 93 return the view key? eg:

```js
var keys = _.uniq(_.compact(summaries.map(function(summary) {
  if (summary.subject && summary.subject.type === 'reference') {
    return ['shortcode', summary.subject.value];
  }
})));
if (!keys.length) {
  return $q.resolve(summaries);
}
```

Reply:
Underscore's uniq uses === operator to compare elements; since the elements would be arrays, I would have to write a similar function which works with array elements - and I thought that was a bit overkill. This was how I wrote it initially :)

Reply:
Ahh that makes sense.